src/validation.js

```diff
@@ -1,11 +1,55 @@
+import { NamedNode } from "rdflib";
 import { FORM, RDF } from "./namespaces.js";
-import { fieldsForForm } from "./fields-for-form.js";
+import { fieldsForForm, fieldsForSubForm } from "./fields-for-form.js";
 import { check, checkTriples } from "./constraints.js";
+import { getScope } from "./get-scope.js";
+import { triplesForScope } from "./triples-for.js";
 
 export function validateForm(form, options) {
-  const fields = fieldsForForm(form, options);
-  const fieldValidations = fields.map((field) => validateField(field, options));
-  return fieldValidations.reduce((acc, value) => acc && value, true);
+  const topLevelFields = fieldsForForm(form, options);
+
+  return validateFields(topLevelFields, options);
+}
+
+function validateFields(fields, options) {
+  const fieldValidations = fields.map((field) => {
+    if (isListing(field, options)) {
+      const listing = field;
+      const scope = getScope(listing, options);
+      const { values: subFormSourceNodes } = triplesForScope(scope, options);
+      const hasSubFormData = subFormSourceNodes.length > 0;
+
+      if (hasSubFormData) {
+        const { store } = options;
+        const subForm = store.any(listing, FORM("each"), undefined);
+        const subformFields = fieldsForSubForm(subForm, options);
+
+        return subFormSourceNodes
+          .map((sourceNode) => {
+            return validateFields(subformFields, { ...options, sourceNode });
+          })
+          .every(Boolean);
+      } else {
+        // TODO: should we validate sh:minCount / sh:maxCount?
```

Comment:
Do we consider this in scope for this PR? I think the current implementation already covers most use-cases assuming the sub form fields also contain validations.